In [ ]:
import pandas as pd

# Load data using relative paths
brands = pd.read_csv('../data/raw/brands.csv')
orderlines = pd.read_csv('../data/raw/orderlines.csv')
orders = pd.read_csv('../data/raw/orders.csv')
products = pd.read_csv('../data/raw/products.csv')

# Create working copies
brands_df = brands.copy()
orderlines_df = orderlines.copy()
orders_df = orders.copy()
products_df = products.copy()

In [2]:
orders_df.shape

(226909, 4)

In [30]:
orders_df.columns

Index(['order_id', 'created_date', 'total_paid', 'state'], dtype='object')

In [3]:
nrows = orders_df.shape[0]
ncols = orders_df.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 226909
The number of columns is 4


In [4]:
orders_df.size

907636

In [5]:
orders_df.shape[0] * orders_df.shape[1] == orders_df.size

True

In [6]:
orders_df.ndim

2

In [9]:
orders_df.head(20)

order_id         created_date  total_paid      state
0     241319  2017-01-02 13:35:40       44.99  Cancelled
1     241423  2017-11-06 13:10:02      136.15  Completed
2     242832  2017-12-31 17:40:03       15.76  Completed
3     243330  2017-02-16 10:59:38       84.98  Completed
4     243784  2017-11-24 13:35:19      157.86  Cancelled
5     245275  2017-06-28 11:35:37      149.00  Completed
6     245595  2017-01-21 12:52:47      112.97  Completed
7     245851  2017-04-04 20:58:21       79.99    Pending
8     245941  2017-01-01 10:32:23      183.52  Completed
9     246018  2017-02-13 23:02:57      211.95  Completed
10    246405  2017-11-24 10:01:27      407.96  Completed
11    247524  2018-01-09 14:21:34      167.98  Completed
12    247643  2017-05-26 11:50:35      153.54  Completed
13    250275  2017-11-19 22:34:54       54.99  Completed
14    251302  2017-05-10 12:27:33      140.99  Completed
15    251688  2017-11-24 23:57:07      505.76  Completed
16    251969  2017-11-21 21:04:12      610.07  Completed
17    252371  2017-02-09 12:31:57       27.98    Pending
18    253220  2017-10-04 19:37:10     1610.00  Completed
19    253306  2017-10-04 14:14:45      497.68  Completed

In [11]:
orders_df.tail(20)

order_id         created_date  total_paid            state
226889    527382  2018-03-14 13:47:43        0.00  Shopping Basket
226890    527383  2018-03-14 13:48:51        9.99  Shopping Basket
226891    527384  2018-03-14 13:49:11      249.99      Place Order
226892    527385  2018-03-14 13:49:21       13.99  Shopping Basket
226893    527386  2018-03-14 13:49:48      161.00      Place Order
226894    527387  2018-03-14 13:50:06        9.99  Shopping Basket
226895    527388  2018-03-14 13:51:59       34.98  Shopping Basket
226896    527389  2018-03-14 13:52:15        9.99  Shopping Basket
226897    527390  2018-03-14 13:52:33        9.99  Shopping Basket
226898    527391  2018-03-14 13:57:54       54.98          Pending
226899    527392  2018-03-14 13:55:11       34.99          Pending
226900    527393  2018-03-14 13:53:57       13.99  Shopping Basket
226901    527394  2018-03-14 13:55:15      161.00      Place Order
226902    527395  2018-03-14 13:55:22       20.99  Shopping Basket
226903    527396  2018-03-14 13:55:44      141.58      Place Order
226904    527397  2018-03-14 13:56:38       42.99      Place Order
226905    527398  2018-03-14 13:57:25       42.99  Shopping Basket
226906    527399  2018-03-14 13:57:34      141.58  Shopping Basket
226907    527400  2018-03-14 13:57:41       19.98  Shopping Basket
226908    527401  2018-03-14 13:58:36       18.98      Place Order

In [22]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [13]:
orders_df.describe()

order_id     total_paid
count  226909.000000  226904.000000
mean   413296.482480     569.225818
std     65919.250331    1761.778002
min    241319.000000       0.000000
25%    356263.000000      34.190000
50%    413040.000000     112.990000
75%    470553.000000     525.980000
max    527401.000000  214747.530000

In [14]:
orders_df.nunique()

order_id        226909
created_date    224828
total_paid       31236
state                5
dtype: int64

In [29]:
orders_df['state'].unique()

array(['Cancelled', 'Completed', 'Pending', 'Shopping Basket',
       'Place Order'], dtype=object)

In [24]:
mask = orders_df['created_date'].duplicated(keep=False)
orders_df.loc[mask,:]

order_id         created_date  total_paid            state
126       290383  2017-01-23 12:29:12      114.98        Completed
155       296253  2017-01-10 11:43:43      308.95        Completed
682       300055  2017-01-01 23:32:12      197.82        Completed
731       300106  2017-01-01 23:32:12       56.98      Place Order
975       300360  2017-01-02 10:52:48     1386.98      Place Order
...          ...                  ...         ...              ...
226679    527172  2018-03-14 12:15:18        9.99  Shopping Basket
226720    527213  2018-03-14 12:48:16       27.98          Pending
226782    527275  2018-03-14 12:48:16        3.99      Place Order
226796    527289  2018-03-14 12:59:47       20.98          Pending
226809    527302  2018-03-14 12:59:47       13.99  Shopping Basket

[3865 rows x 4 columns]

In [25]:
orders_df.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [28]:
mask = orders_df['total_paid'].isna()
orders_df.loc[mask,:]


order_id         created_date  total_paid    state
127701    427314  2017-11-20 18:54:39         NaN  Pending
132013    431655  2017-11-22 12:15:24         NaN  Pending
147316    447411  2017-11-27 10:32:37         NaN  Pending
148833    448966  2017-11-27 18:54:15         NaN  Pending
149434    449596  2017-11-27 21:52:08         NaN  Pending

In [34]:
orders_df['total_paid'].nsmallest(100)

150     0.0
264     0.0
308     0.0
377     0.0
380     0.0
       ... 
1565    0.0
1575    0.0
1576    0.0
1594    0.0
1599    0.0
Name: total_paid, Length: 100, dtype: float64

In [36]:
mask = orders_df['total_paid'] == 0
orders_df.loc[mask,:]['state'].value_counts()

state
Shopping Basket    9858
Place Order        9419
Cancelled            35
Completed            16
Pending               6
Name: count, dtype: int64

In [37]:
orders_df['total_paid'].nlargest(50)

96028     214747.53
2478      214746.63
183927    214714.31
61659     214642.47
40557     214517.58
32296     159989.83
188457    121521.85
59492     117633.68
67674     101887.33
59208     100738.45
34826      99862.03
164560     99648.10
222959     99515.06
35280      99142.98
131309     99069.47
49870      97250.01
33492      92094.00
77304      90989.60
113917     90077.94
111005     87825.33
62310      87020.97
35319      79981.00
157910     61363.35
12474      59499.63
82807      58300.01
101844     57083.91
54552      56396.00
46728      55930.71
18714      49989.90
53258      48684.97
37449      47439.96
31810      44714.85
131993     42988.90
29827      42476.54
82140      41853.23
76075      41509.72
49640      39399.84
216804     37783.03
200061     35930.10
215170     35813.54
211222     34625.57
218280     32850.07
166881     32068.59
106101     30663.94
54210      30051.91
40151      29749.76
92110      27546.04
61707      27179.91
774        26924.85
122982     26530.84


## Documentation

### Dataset Overview
- **226,909 rows** and **4 columns:** `order_id`, `created_date`, `total_paid`, `state`
- `order_id` has no duplicates
- `state` contains these values: 'Cancelled', 'Completed', 'Pending', 'Shopping Basket', 'Place Order'

### Data Quality Notes
- **5 null values** in `total_paid` - all with status 'Pending'
- `total_paid` is often 0 (~20,000 cases) which indicates a potential data issue:
  - Shopping Basket: 9,858
  - Place Order: 9,419
  - Cancelled: 35
  - Completed: 16
  - Pending: 6

### Summary Statistics (without data cleaning)
- **Average order value:** $569.23 USD
- **Median:** $112.99 USD
- **75th percentile:** $525.98 (three quarters of all orders are below this value)
- **10 orders** exceeded $100,000 USD